nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py

```diff
@@ -0,0 +1,267 @@
+            confidence_method_cfg: config for the confidence
+        """
+        super().__init__()
+        self.decoder = decoder
+        self.joint = joint
+        self.durations: list[int] = list(durations)
+        self._blank_index = blank_index
+        self.max_symbols = max_symbols_per_step
+        self.preserve_alignments = preserve_alignments
+        self.preserve_frame_confidence = preserve_frame_confidence
+        self._SOS = self._blank_index
+        self._init_confidence_method(confidence_method_cfg=confidence_method_cfg)
+        assert self._SOS == self._blank_index  # "blank as pad" algorithm only
+
+    def __call__(
+        self, x: torch.Tensor, out_len: torch.Tensor,
+    ) -> Tuple[rnnt_utils.BatchedHyps, Optional[rnnt_utils.BatchedAlignments], Any]:
+        """
+        Optimized batched greedy decoding.
+        Iterates over labels, on each step finding the next non-blank label
+        (evaluating Joint multiple times in inner loop); It uses a minimal possible amount of calls
+        to prediction network (with maximum possible batch size),
+        which makes it especially useful for scaling the prediction network.
+        During decoding all active hypotheses ("texts") have the same lengths.
+
+        Args:
+            x: output from the encoder
+            out_len: lengths of the utterances in `x`
+        """
+        batch_size, max_time, _unused = x.shape
+        device = x.device
+
+        x = self.joint.project_encoder(x)  # do not recalculate joint projection, project only once
+
+        # init output structures: BatchedHyps (for results), BatchedAlignments + last decoder state
+        # init empty batched hypotheses
+        batched_hyps = rnnt_utils.BatchedHyps(
+            batch_size=batch_size,
+            init_length=max_time * self.max_symbols if self.max_symbols is not None else max_time,
+            device=x.device,
+            float_dtype=x.dtype,
+        )
+        # sample state, will be replaced further when the decoding for hypothesis is done
+        last_decoder_state = self.decoder.initialize_state(x)
+        # init alignments if necessary
+        use_alignments = self.preserve_alignments or self.preserve_frame_confidence
+        # always use alignments variable - for torch.jit adaptation, but keep it as minimal as possible
+        alignments = rnnt_utils.BatchedAlignments(
+            batch_size=batch_size,
+            logits_dim=self.joint.num_classes_with_blank,
+            init_length=max_time * 2 if use_alignments else 1,  # blank for each timestep + text tokens
+            device=x.device,
+            float_dtype=x.dtype,
+            store_alignments=self.preserve_alignments,
+            store_frame_confidence=self.preserve_frame_confidence,
+        )
+
+        # durations
+        duration_lengths = len(self.durations)
+        all_durations = torch.tensor(self.durations, device=device).to(torch.long)
```

Comment:
Can we move this as a member of the class so that it doesn't need to be initialized every time?

Reply:
I renamed  `duration_lengths` -> `num_durations`, and used tensor for `self.durations`. I agree there is no need to convert it to tensor here every time.
I think `num_durations` introduces no overhead for now, and I would prefer not to add denormalization to the class.
`self.durations` are stored on the CPU and moved to the other device only on demand. It is a safer version with mostly zero overhead (asynchronously moving several numbers to GPU is a lightweight operation compared to overall decoding). If we move the class member to GPU (e.g., `self.durations = self.durations.to(...)`), there will be undesirable side-effects in the multi-gpu environment (even if we convert this class to `nn.Module`, it is used through a "proxy" class that is not an `nn.Module`), e.g., `model.to("cuda:0"); model.transcribe(...); model.to("cuda:1")` – the tensor will be still on "cuda:0". I added a comment in the code.

Reply:
Ah OK, this makes sense. Thanks for the clarification.